File: backend/app/schemas/company.py
"""Public request and response contracts for SEC company research."""

from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class CompanySearchParams(BaseModel):
    """Validated company-search query parameters."""

    model_config = ConfigDict(extra="forbid", frozen=True, str_strip_whitespace=True)

    query: str = Field(min_length=2, max_length=100)
    limit: int = Field(default=8, ge=1, le=20)


class CompanyFilingsParams(BaseModel):
    """Validated filing-filter query parameters."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    forms: tuple[str, ...] = Field(min_length=1, max_length=10)
    limit: int = Field(default=20, ge=1, le=100)

    @field_validator("forms")
    @classmethod
    def forms_must_be_unique(cls, forms: tuple[str, ...]) -> tuple[str, ...]:
        if len(set(forms)) != len(forms):
            raise ValueError("forms must not contain duplicates")
        return forms


class ErrorDetail(BaseModel):
    """Stable error information safe to expose to API consumers."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    code: str
    message: str


class ErrorResponse(BaseModel):
    """Envelope used by predictable application errors."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    error: ErrorDetail


class CompanySearchResult(BaseModel):
    """Normalized SEC ticker mapping entry."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    ticker: str
    company_name: str
    cik: str = Field(pattern=r"^\d{10}$")


class CompanySearchResponse(BaseModel):
    """Ranked company-search results."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    query: str
    results: list[CompanySearchResult]


class CompanyAddress(BaseModel):
    """Normalized SEC business or mailing address."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    street1: str | None = None
    street2: str | None = None
    city: str | None = None
    state_or_country: str | None = None
    state_or_country_description: str | None = None
    postal_code: str | None = None


class FormerCompanyName(BaseModel):
    """A former registrant name and the period in which it applied."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    name: str
    from_date: date | None = None
    to_date: date | None = None


class CompanyProfileResponse(BaseModel):
    """Normalized SEC company submissions metadata."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    ticker: str
    company_name: str
    cik: str = Field(pattern=r"^\d{10}$")
    sic_code: str | None = None
    sic_description: str | None = None
    exchanges: list[str]
    fiscal_year_end: str | None = None
    state_of_incorporation: str | None = None
    business_address: CompanyAddress | None = None
    mailing_address: CompanyAddress | None = None
    former_names: list[FormerCompanyName]
    sec_company_url: str


class FilingRecord(BaseModel):
    """One normalized recent filing from SEC submissions data."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    accession_number: str
    form: str
    filing_date: date
    report_date: date | None = None
    acceptance_timestamp: datetime | None = None
    primary_document: str
    filing_detail_url: str
    primary_document_url: str
    description: str | None = None
    items: str | None = None


class CompanyFilingsResponse(BaseModel):
    """Recent filings for one company and the applied form filters."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    ticker: str
    cik: str = Field(pattern=r"^\d{10}$")
    forms: list[str]
    filings: list[FilingRecord]

File: frontend/src/types/company.ts
export interface CompanySearchResult {
  ticker: string;
  company_name: string;
  cik: string;
}

export interface CompanySearchResponse {
  query: string;
  results: CompanySearchResult[];
}

export interface CompanyAddress {
  street1: string | null;
  street2: string | null;
  city: string | null;
  state_or_country: string | null;
  state_or_country_description: string | null;
  postal_code: string | null;
}

export interface FormerCompanyName {
  name: string;
  from_date: string | null;
  to_date: string | null;
}

export interface CompanyProfileResponse {
  ticker: string;
  company_name: string;
  cik: string;
  sic_code: string | null;
  sic_description: string | null;
  exchanges: string[];
  fiscal_year_end: string | null;
  state_of_incorporation: string | null;
  business_address: CompanyAddress | null;
  mailing_address: CompanyAddress | null;
  former_names: FormerCompanyName[];
  sec_company_url: string;
}

export interface FilingRecord {
  accession_number: string;
  form: string;
  filing_date: string;
  report_date: string | null;
  acceptance_timestamp: string | null;
  primary_document: string;
  filing_detail_url: string;
  primary_document_url: string;
  description: string | null;
  items: string | null;
}

export interface CompanyFilingsResponse {
  ticker: string;
  cik: string;
  forms: string[];
  filings: FilingRecord[];
}

export type FundamentalMetricKey =
  | "revenue"
  | "operating_income"
  | "net_income"
  | "diluted_eps"
  | "cash"
  | "debt"
  | "operating_margin"
  | "net_margin";

export type FundamentalPeriod = "annual" | "quarterly";

export interface FundamentalComponentSource {
  metric_key: string;
  value: number;
  unit: string;
  taxonomy: string;
  source_tag: string;
  accession_number: string;
  source_filing_url: string;
}

export interface FundamentalFact {
  metric_key: FundamentalMetricKey;
  value: number;
  unit: string;
  period_start: string | null;
  period_end: string;
  fiscal_year: number;
  fiscal_period: string;
  form: string;
  filed_date: string;
  accession_number: string;
  frame: string | null;
  taxonomy: string;
  source_tag: string;
  is_fallback: boolean;
  is_derived: boolean;
  is_restated: boolean;
  source_filing_url: string;
  component_sources: FundamentalComponentSource[];
}

export interface FundamentalMetricSeries {
  metric_key: FundamentalMetricKey;
  label: string;
  unit: string;
  period: FundamentalPeriod;
  facts: FundamentalFact[];
}

export interface LatestFundamentalValue {
  metric_key: FundamentalMetricKey;
  label: string;
  fact: FundamentalFact;
}

export interface DataQualityWarning {
  code: string;
  message: string;
  metric_key: FundamentalMetricKey | null;
}

export interface UnavailableMetric {
  metric_key: FundamentalMetricKey;
  label: string;
  period: FundamentalPeriod;
  reason: string;
}

export interface CompanyFundamentalsResponse {
  company: {
    ticker: string;
    company_name: string;
    cik: string;
  };
  data_as_of: string;
  annual: FundamentalMetricSeries[];
  quarterly: FundamentalMetricSeries[];
  latest_values: LatestFundamentalValue[];
  warnings: DataQualityWarning[];
  unavailable_metrics: UnavailableMetric[];
  provenance: {
    provider: "SEC EDGAR Company Facts";
    company_facts_url: string;
  };
}
